### PR TITLE
Improve sparsity pattern

### PR DESCRIPTION
**Pull request by chrisrichardson**
* Simplifies internals, by caching owned and unowned _rows_ separately (instead of _columns_)
* Use int32 indices for off-diagonal (unowned columns)
* Creates a new IndexMap for columns after assembly (with full local-to-global map)

At the moment, PETSc Mat just uses nnz per row to initialise, but this change will make it easier to use other libraries which need the full graph to initialise their matrix structures.

### CODE REVIEW

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 1]
```diff
@@ -212,27 +165,20 @@ void SparsityPattern::insert(const tcb::span<const std::int32_t>& rows,
...
-      {
-        if (cols[j] < local_size1)
-          _diagonal_cache[rows[i]].push_back(cols[j]);
-        else
-        {
-          _off_diagonal_cache[rows[i]].push_back(
-              ghosts1[cols[j] - local_size1]);
-        }
-      }
+      for (std::int32_t col : cols)
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:175]: Could use `insert` here.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 2]
```diff
@@ -212,27 +165,20 @@ void SparsityPattern::insert(const tcb::span<const std::int32_t>& rows,
...
+        _cache_owned[row].push_back(col);
+    }
+    else if (row < size0)
+    {
+      for (std::int32_t col : cols)
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:180]: Could use `insert` here.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 3]
```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
...
       }
       else
       {
-        assert(row_local < (std::int32_t)_off_diagonal_cache.size());
-        _off_diagonal_cache[row_local].push_back(col);
+        // Column index may not exist in column indexmap
+        auto it = global_to_local.insert({col, local_i});
+        if (it.second)
+        {
+          ++local_i;
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:314]: Would be clearer if this as remove to the bottom of the code block.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 4]
```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
...
-    row.erase(std::unique(row.begin(), row.end()), row.end());
+    const std::vector<std::int32_t>::iterator it_end
+        = std::unique(row.begin(), row.end());
+
+    // Find position of first "off-diagonal" column
+    const std::vector<std::int32_t>::iterator it_diag
+        = std::lower_bound(row.begin(), it_end, local_size1);
+
+    adj_data.insert(adj_data.end(), row.begin(), it_diag);
+    adj_counts[i] += (it_diag - row.begin());
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:341]: Use `std::distance`.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 5]
```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
...
+    adj_data_off.insert(adj_data_off.end(), it_diag, it_end);
+    adj_counts_off[i] += (it_end - it_diag);
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:343]: Use `std::distance`.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 6]
```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
...
-      _off_diagonal_cache);
-  std::vector<std::vector<std::int64_t>>().swap(_off_diagonal_cache);
+  std::vector<std::int32_t> adj_offsets(local_size0 + 1);
+  std::partial_sum(adj_counts.begin(), adj_counts.end(),
+                   adj_offsets.begin() + 1);
+  std::vector<std::int32_t> adj_offsets_off(local_size0 + 1);
+  std::partial_sum(adj_counts_off.begin(), adj_counts_off.end(),
+                   adj_offsets_off.begin() + 1);
+
+  // FIXME: after transferring rows, the "correct" row map has no ghosts,
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:355]: This is a bit cryptic. What's the "original" l2g? I would expect this is just one.

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 7]
```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
...
+  // but PETSc Mat needs to know original L2G map.
+  // _index_maps[0] = std::make_shared<common::IndexMap>(comm, local_size0);
+  _diagonal = std::make_shared<graph::AdjacencyList<std::int32_t>>(
+      std::move(adj_data), std::move(adj_offsets));
+
+  LOG(INFO) << "Column ghost size increased from "
```
Comment (garth-wells) [inline cpp/dolfinx/la/SparsityPattern.cpp:361]: Why has it increased? Add reason to log comment.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/la/SparsityPattern.cpp (hunk 7 above)
Comment (chrisrichardson) [reply cpp/dolfinx/la/SparsityPattern.cpp:361]: Due to received new columns from other processes - this is noted earlier in the code.

File: cpp/dolfinx/la/SparsityPattern.cpp (hunk 6 above)
Comment (chrisrichardson) [reply cpp/dolfinx/la/SparsityPattern.cpp:355]: I'll expand the explanation here

**Pull request merged**